### PR TITLE
Add arith-expand pass to lower ceildiv, floordiv ops

### PR DESCRIPTION
**Pull request by harsh-nod**
This PR adds the arith expand pass that decomposes ops like ceildiv and floordiv into primitive arith ops that can be lowered to LLVM.

### CODE REVIEW

**Review by kuhar — COMMENTED** [review]
Could we add a test that requires this change to properly lower to llvm?

**Comment by harsh-nod** [comment]
> Could we add a test that requires this change to properly lower to llvm?

Sure, do you know what the command line flags would be to pass to iree-opt to get it to run the lower to llvmgpu passes? I tried "iree-opt --split-input-file --iree-gpu-test-target=gfx940 --pass-pipeline="builtin.module(hal.executable(hal.executable.variant(builtin.module(iree-codegen-llvmgpu-configuration-pipeline), iree-codegen-linalg-to-rocdl-pipeline)))" but that doesn't work.

**Comment by kuhar** [comment]
I think it should be `addLowerToLLVMGPUPasses` -> `buildLLVMGPUCodegenPassPipeline` -> `buildTranslationPassPipeline` -> `--iree-hal-translate-target-executable-variants`

**Comment by kuhar** [comment]
Also seems like this breaks the nvvm pipeline: https://github.com/iree-org/iree/actions/runs/11909531194/job/33187149103?pr=19200#step:7:3311

**Comment by harsh-nod** [comment]
> Also seems like this breaks the nvvm pipeline: https://github.com/iree-org/iree/actions/runs/11909531194/job/33187149103?pr=19200#step:7:3311

Thanks. Looks like that is because the expandArith pass brings in a lot of extra passes. I have uploaded a new version which just populates the patterns for floordiv and ceildiv for rocdl and added a test as well.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/ConvertToROCDL.cpp
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/ConvertToROCDL.cpp:155]: I wonder if we should add these patterns to `populateArithToAMDGPUConversionPatterns`

**Review by kuhar — APPROVED** [review]
LGTM but maybe lets wait for a confirmation from @krzysz00

File: compiler/src/iree/compiler/Codegen/LLVMGPU/ConvertToROCDL.cpp
Comment (krzysz00) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/ConvertToROCDL.cpp:155]: I don't think `ArithToAMDGPu` creates any of these ops.

Where are these non-typical divides coming from?

Comment (harsh-nod) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/ConvertToROCDL.cpp:155]: We are generating them in iree-turbine, here specifically: https://github.com/iree-org/iree-turbine/blob/63aaf873973aef48e829712a9e1c849cbe000406/iree/turbine/kernel/wave/codegen.py#L323

Comment (harsh-nod) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/ConvertToROCDL.cpp:155]: But @krzysz00 if you know a better way to do ceildiv on CDNA, please let me know.

**Comment by harsh-nod** [comment]
Also looking at the failures ,

```
_ IREE compile and run: test_center_crop_pad_crop::model.mlir::model.mlir::gpu_rocm_rdna3 _
[gw0] linux -- Python 3.11.9 /home/esaimana/actions-runner/_work/iree/iree/venv/bin/python
[XPASS(strict)] Expected compilation to fail (included in 'expected_compile_failures')
_ IREE compile and run: test_center_crop_pad_crop_and_pad::model.mlir::model.mlir::gpu_rocm_rdna3 _
[gw0] linux -- Python 3.11.9 /home/esaimana/actions-runner/_work/iree/iree/venv/bin/python
[XPASS(strict)] Expected compilation to fail (included in 'expected_compile_failures')
_ IREE compile and run: test_center_crop_pad_crop_axes_chw::model.mlir::model.mlir::gpu_rocm_rdna3 _
[gw0] linux -- Python 3.11.9 /home/esaimana/actions-runner/_work/iree/iree/venv/bin/python
[XPASS(strict)] Expected compilation to fail (included in 'expected_compile_failures')
_ IREE compile and run: test_center_crop_pad_crop_axes_hwc::model.mlir::model.mlir::gpu_rocm_rdna3 _
[gw0] linux -- Python 3.11.9 /home/esaimana/actions-runner/_work/iree/iree/venv/bin/python
[XPASS(strict)] Expected compilation to fail (included in 'expected_compile_failures')
_ IREE compile and run: test_center_crop_pad_crop_negative_axes_hwc::model.mlir::model.mlir::gpu_rocm_rdna3 _
[gw0] linux -- Python 3.11.9 /home/esaimana/actions-runner/_work/iree/iree/venv/bin/python
[XPASS(strict)] Expected compilation to fail (included in 'expected_compile_failures')
_ IREE compile and run: test_center_crop_pad_pad::model.mlir::model.mlir::gpu_rocm_rdna3 _
[gw0] linux -- Python 3.11.9 /home/esaimana/actions-runner/_work/iree/iree/venv/bin/python
[XPASS(strict)] Expected compilation to fail (included in 'expected_compile_failures')
```

am I correct in reading this as this PR is making the above tests that failed compilation , compile successfully?

**Comment by ScottTodd** [comment]
> Also looking at the failures ,
> 
> ```
> _ IREE compile and run: test_center_crop_pad_crop::model.mlir::model.mlir::gpu_rocm_rdna3 _
> [gw0] linux -- Python 3.11.9 /home/esaimana/actions-runner/_work/iree/iree/venv/bin/python
> [XPASS(strict)] Expected compilation to fail (included in 'expected_compile_failures')
> _ IREE compile and run: test_center_crop_pad_crop_and_pad::model.mlir::model.mlir::gpu_rocm_rdna3 _
> [gw0] linux -- Python 3.11.9 /home/esaimana/actions-runner/_work/iree/iree/venv/bin/python
> [XPASS(strict)] Expected compilation to fail (included in 'expected_compile_failures')
> _ IREE compile and run: test_center_crop_pad_crop_axes_chw::model.mlir::model.mlir::gpu_rocm_rdna3 _
> [gw0] linux -- Python 3.11.9 /home/esaimana/actions-runner/_work/iree/iree/venv/bin/python
> [XPASS(strict)] Expected compilation to fail (included in 'expected_compile_failures')
> _ IREE compile and run: test_center_crop_pad_crop_axes_hwc::model.mlir::model.mlir::gpu_rocm_rdna3 _
> [gw0] linux -- Python 3.11.9 /home/esaimana/actions-runner/_work/iree/iree/venv/bin/python
> [XPASS(strict)] Expected compilation to fail (included in 'expected_compile_failures')
> _ IREE compile and run: test_center_crop_pad_crop_negative_axes_hwc::model.mlir::model.mlir::gpu_rocm_rdna3 _
> [gw0] linux -- Python 3.11.9 /home/esaimana/actions-runner/_work/iree/iree/venv/bin/python
> [XPASS(strict)] Expected compilation to fail (included in 'expected_compile_failures')
> _ IREE compile and run: test_center_crop_pad_pad::model.mlir::model.mlir::gpu_rocm_rdna3 _
> [gw0] linux -- Python 3.11.9 /home/esaimana/actions-runner/_work/iree/iree/venv/bin/python
> [XPASS(strict)] Expected compilation to fail (included in 'expected_compile_failures')
> ```
> 
> am I correct in reading this as this PR is making the above tests that failed compilation , compile successfully?

Yes. You can download the .json file from the bottom of https://github.com/iree-org/iree/actions/runs/11920404022?pr=19200 and add it to https://github.com/iree-org/iree/tree/main/tests/external/iree-test-suites/onnx_ops to accept the new test results.

**Comment by harsh-nod** [comment]
> > Also looking at the failures ,
> > ```
> > _ IREE compile and run: test_center_crop_pad_crop::model.mlir::model.mlir::gpu_rocm_rdna3 _
> > [gw0] linux -- Python 3.11.9 /home/esaimana/actions-runner/_work/iree/iree/venv/bin/python
> > [XPASS(strict)] Expected compilation to fail (included in 'expected_compile_failures')
> > _ IREE compile and run: test_center_crop_pad_crop_and_pad::model.mlir::model.mlir::gpu_rocm_rdna3 _
> > [gw0] linux -- Python 3.11.9 /home/esaimana/actions-runner/_work/iree/iree/venv/bin/python
> > [XPASS(strict)] Expected compilation to fail (included in 'expected_compile_failures')
> > _ IREE compile and run: test_center_crop_pad_crop_axes_chw::model.mlir::model.mlir::gpu_rocm_rdna3 _
> > [gw0] linux -- Python 3.11.9 /home/esaimana/actions-runner/_work/iree/iree/venv/bin/python
> > [XPASS(strict)] Expected compilation to fail (included in 'expected_compile_failures')
> > _ IREE compile and run: test_center_crop_pad_crop_axes_hwc::model.mlir::model.mlir::gpu_rocm_rdna3 _
> > [gw0] linux -- Python 3.11.9 /home/esaimana/actions-runner/_work/iree/iree/venv/bin/python
> > [XPASS(strict)] Expected compilation to fail (included in 'expected_compile_failures')
> > _ IREE compile and run: test_center_crop_pad_crop_negative_axes_hwc::model.mlir::model.mlir::gpu_rocm_rdna3 _
> > [gw0] linux -- Python 3.11.9 /home/esaimana/actions-runner/_work/iree/iree/venv/bin/python
> > [XPASS(strict)] Expected compilation to fail (included in 'expected_compile_failures')
> > _ IREE compile and run: test_center_crop_pad_pad::model.mlir::model.mlir::gpu_rocm_rdna3 _
> > [gw0] linux -- Python 3.11.9 /home/esaimana/actions-runner/_work/iree/iree/venv/bin/python
> > [XPASS(strict)] Expected compilation to fail (included in 'expected_compile_failures')
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > am I correct in reading this as this PR is making the above tests that failed compilation , compile successfully?
> 
> Yes. You can download the .json file from the bottom of https://github.com/iree-org/iree/actions/runs/11920404022?pr=19200 and add it to https://github.com/iree-org/iree/tree/main/tests/external/iree-test-suites/onnx_ops to accept the new test results.

Thanks, will do that once I have approval from @krzysz00 .

File: compiler/src/iree/compiler/Codegen/LLVMGPU/ConvertToROCDL.cpp
Comment (krzysz00) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/ConvertToROCDL.cpp:155]: Nah, should be good

**Review by krzysz00 — APPROVED** [review]
Approved, though it might be a good idea to add this to all the other LLVM backends - there's nothing CDNA-specific here

**Comment by harsh-nod** [comment]
> Approved, though it might be a good idea to add this to all the other LLVM backends - there's nothing CDNA-specific here

Sounds good. I will add it to convert_to_nvvm as well before I merge it in.

**Comment by harsh-nod** [comment]
> > Also looking at the failures ,
> > ```
> > _ IREE compile and run: test_center_crop_pad_crop::model.mlir::model.mlir::gpu_rocm_rdna3 _
> > [gw0] linux -- Python 3.11.9 /home/esaimana/actions-runner/_work/iree/iree/venv/bin/python
> > [XPASS(strict)] Expected compilation to fail (included in 'expected_compile_failures')
> > _ IREE compile and run: test_center_crop_pad_crop_and_pad::model.mlir::model.mlir::gpu_rocm_rdna3 _
> > [gw0] linux -- Python 3.11.9 /home/esaimana/actions-runner/_work/iree/iree/venv/bin/python
> > [XPASS(strict)] Expected compilation to fail (included in 'expected_compile_failures')
> > _ IREE compile and run: test_center_crop_pad_crop_axes_chw::model.mlir::model.mlir::gpu_rocm_rdna3 _
> > [gw0] linux -- Python 3.11.9 /home/esaimana/actions-runner/_work/iree/iree/venv/bin/python
> > [XPASS(strict)] Expected compilation to fail (included in 'expected_compile_failures')
> > _ IREE compile and run: test_center_crop_pad_crop_axes_hwc::model.mlir::model.mlir::gpu_rocm_rdna3 _
> > [gw0] linux -- Python 3.11.9 /home/esaimana/actions-runner/_work/iree/iree/venv/bin/python
> > [XPASS(strict)] Expected compilation to fail (included in 'expected_compile_failures')
> > _ IREE compile and run: test_center_crop_pad_crop_negative_axes_hwc::model.mlir::model.mlir::gpu_rocm_rdna3 _
> > [gw0] linux -- Python 3.11.9 /home/esaimana/actions-runner/_work/iree/iree/venv/bin/python
> > [XPASS(strict)] Expected compilation to fail (included in 'expected_compile_failures')
> > _ IREE compile and run: test_center_crop_pad_pad::model.mlir::model.mlir::gpu_rocm_rdna3 _
> > [gw0] linux -- Python 3.11.9 /home/esaimana/actions-runner/_work/iree/iree/venv/bin/python
> > [XPASS(strict)] Expected compilation to fail (included in 'expected_compile_failures')
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > am I correct in reading this as this PR is making the above tests that failed compilation , compile successfully?
> 
> Yes. You can download the .json file from the bottom of https://github.com/iree-org/iree/actions/runs/11920404022?pr=19200 and add it to https://github.com/iree-org/iree/tree/main/tests/external/iree-test-suites/onnx_ops to accept the new test results.

@ScottTodd - so what should be the order in which I do these? Do I first merge the PR into iree and then the PR into tree-test-suites or the other way around?

**Comment by ScottTodd** [comment]
> > > Also looking at the failures ,
> > > ```
> > > _ IREE compile and run: test_center_crop_pad_crop::model.mlir::model.mlir::gpu_rocm_rdna3 _
> > > [gw0] linux -- Python 3.11.9 /home/esaimana/actions-runner/_work/iree/iree/venv/bin/python
> > > [XPASS(strict)] Expected compilation to fail (included in 'expected_compile_failures')
> > > _ IREE compile and run: test_center_crop_pad_crop_and_pad::model.mlir::model.mlir::gpu_rocm_rdna3 _
> > > [gw0] linux -- Python 3.11.9 /home/esaimana/actions-runner/_work/iree/iree/venv/bin/python
> > > [XPASS(strict)] Expected compilation to fail (included in 'expected_compile_failures')
> > > _ IREE compile and run: test_center_crop_pad_crop_axes_chw::model.mlir::model.mlir::gpu_rocm_rdna3 _
> > > [gw0] linux -- Python 3.11.9 /home/esaimana/actions-runner/_work/iree/iree/venv/bin/python
> > > [XPASS(strict)] Expected compilation to fail (included in 'expected_compile_failures')
> > > _ IREE compile and run: test_center_crop_pad_crop_axes_hwc::model.mlir::model.mlir::gpu_rocm_rdna3 _
> > > [gw0] linux -- Python 3.11.9 /home/esaimana/actions-runner/_work/iree/iree/venv/bin/python
> > > [XPASS(strict)] Expected compilation to fail (included in 'expected_compile_failures')
> > > _ IREE compile and run: test_center_crop_pad_crop_negative_axes_hwc::model.mlir::model.mlir::gpu_rocm_rdna3 _
> > > [gw0] linux -- Python 3.11.9 /home/esaimana/actions-runner/_work/iree/iree/venv/bin/python
> > > [XPASS(strict)] Expected compilation to fail (included in 'expected_compile_failures')
> > > _ IREE compile and run: test_center_crop_pad_pad::model.mlir::model.mlir::gpu_rocm_rdna3 _
> > > [gw0] linux -- Python 3.11.9 /home/esaimana/actions-runner/_work/iree/iree/venv/bin/python
> > > [XPASS(strict)] Expected compilation to fail (included in 'expected_compile_failures')
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > am I correct in reading this as this PR is making the above tests that failed compilation , compile successfully?
> > 
> > 
> > Yes. You can download the .json file from the bottom of https://github.com/iree-org/iree/actions/runs/11920404022?pr=19200 and add it to https://github.com/iree-org/iree/tree/main/tests/external/iree-test-suites/onnx_ops to accept the new test results.
> 
> @ScottTodd - so what should be the order in which I do these? Do I first merge the PR into iree and then the PR into tree-test-suites or the other way around?

No changes needed in the iree-test-suites repo. The test statuses are entirely in this repo, in the folder I linked.

**Comment by ScottTodd** [comment]
(docs for those tests: https://iree.dev/developers/general/testing-guide/#onnx-operator-tests)

**Comment by harsh-nod** [comment]
> (docs for those tests: https://iree.dev/developers/general/testing-guide/#onnx-operator-tests)

Thanks, that helps.

**Pull request merged**
